src/pyhf/constraints.py

```diff
@@ -1,33 +1,39 @@
 from . import get_backend, default_backend
 from . import events
+from .paramview import ParamViewer
 
 
 class gaussian_constraint_combined(object):
-    def __init__(self, pdfconfig):
+    def __init__(self, pdfconfig, batch_size=None):
+        self.batch_size = batch_size
         # iterate over all constraints order doesn't matter....
 
-        self.par_indices = list(range(len(pdfconfig.suggested_init())))
         self.data_indices = list(range(len(pdfconfig.auxdata)))
         self.parset_and_slice = [
-            (pdfconfig.param_set(cname), pdfconfig.par_slice(cname))
+            pdfconfig.param_set(cname) for cname in pdfconfig.auxdata_order
+        ]
+
+        pnames = [
+            cname
             for cname in pdfconfig.auxdata_order
+            if pdfconfig.param_set(cname).pdf_type == 'normal'
         ]
-        self._precompute()
-        events.subscribe('tensorlib_changed')(self._precompute)
 
-    def _precompute(self):
-        tensorlib, _ = get_backend()
+        parfield_shape = (self.batch_size or 1, len(pdfconfig.suggested_init()))
+        self.parameter_helper = ParamViewer(parfield_shape, pdfconfig.par_map, pnames)
+
         start_index = 0
         normal_constraint_data = []
-        normal_constraint_mean_indices = []
         normal_constraint_sigmas = []
-        for parset, parslice in self.parset_and_slice:
+        for parset in self.parset_and_slice:
```

Comment:
`self.parset_and_slice` should just become `self.parsets`

Reply:
done